### ISSUE TITLE
Performance/genericReset: Make {.noInit.} an unnecessary optimization for heap types

### ISSUE THREAD

**Issue by mratsim** [nim-lang/Nim#8745]
# Context

Nim refcounting GC introduces several useless `genericReset` and `genericResetAux` with very high performance impact on common code. GC mark-and-sweep does not have this issue (tested on completely-unscientific-benchmark snippet).

The workaround is using `{.noInit.}` which is undesirable.

The impact is that Nim code using ref types, including seq and maybe strings too, is slow.

The code snippets below will show from **30% to 6x performance loss** on naive code vs `{.noInit.}` code

This will show on many benchmarks, especially when prefer to use "default, stable and idiomatic language constructs" like [here](https://github.com/frol/completely-unscientific-benchmarks/pull/17#issuecomment-388939830).

# Related

This is strongly related to:
* Assignment in Arraymancer woes - https://irclogs.nim-lang.org/12-10-2017.html#19:04:46 and a couple older discussion:
  ```Nim
  template newTensorUninit*[T](shape: varargs[int], result: var Tensor[T])=
    # No genericReset
    tensorCpu(shape, result)
    result.data = newSeqUninit[T](result.size)

  proc newTensorUninit*[T](shape: varargs[int]): Tensor[T] {.noSideEffect, inline.} =
    # genericReset
    tensorCpu(shape, result)
    result.data = newSeqUninit[T](result.size)
  ```
* #6501 - Assignment in result variables generates genericResets
    * The difference is that #6501 is about `result = newFoo()` while this is also about `result.field1 = something()` (partial instantiation)
* #7443 - 3x performance regression in Arraymancer benchmarks
    * This was solved by https://github.com/nim-lang/Nim/commit/88cf6573e04bd7ee8762aa336460b9748f0d4644
    * `{.noInit.}` was not taken into account anymore
    * This is regression was introduced by https://github.com/nim-lang/Nim/commit/358709e9cbc4ed3e084dc6d1db313a862b8356e4 which was solving
       a `genericReset` related codegenbug https://github.com/nim-lang/Nim/issues/6960
* https://github.com/frol/completely-unscientific-benchmarks/pull/17 - Nim completely unscientific benchmark
    * https://irclogs.nim-lang.org/14-05-2018.html#19:31:01 - @Araq:
         > gosh just fix that genericReset
    * https://irclogs.nim-lang.org/15-05-2018.html#08:34:11 - @PMunch:
         > That makes a huge difference
    * https://irclogs.nim-lang.org/21-08-2018.html#09:34:42 - @Araq:
         > I don't want programmer to write .noInit

# What introduces genericReset:

This resetLoc proc on line 741

https://github.com/nim-lang/Nim/blob/55a86497491586dc1b5ad2f6303d53992b999ab5/compiler/cgen.nim#L715-L744

# Idea

Can we piggyback on `not nil` for ref return types and completely remove that for stack objects?

# Snippets with perf ratio

1. **2.75x perf ratio** - Object containing a seq that is partially initialized
    ```Nim
    import times

    type
      MySeq = object
        foo: bool
        bar: int
        baz: array[4, int]
        data: seq[int]

    proc newMySeq(size: int): MySeq {.noInit.}=
      result.data = newSeq[int](size)

    proc head(x: MySeq, n: int): MySeq {.noInit.}=
      result.data = x.data[0 ..< n]

    proc main() =
      const iter = 10_000_000
      when defined(release):
        const report = "Release mode, time taken for " & $iter & " iterations: "
      else:
        const report = "Debug mode, time taken " & $iter & " iterations: "

      let start = cpuTime()
      for _ in 0 ..< iter:
        var x = newMySeq(10)
        x.data[0] = 123
        let y = x.head(3)
        discard y
      let stop = cpuTime()

      echo report & $(stop - start) & " seconds."

    main()

    ## On i5-5227U (Broadwell mobile dual-core  2.7 Ghz turbo 3.1)
    # Release mode, time taken for 10000000 iterations: 2.188145 seconds.

    # With noInit
    # Release mode, time taken for 10000000 iterations: 0.794412 seconds.
    ```

2. **1.31x perf ratio** - Tuple containing seqs that are discarded
   ```Nim
    import times

    type
      MySeqs = tuple[first, second: seq[int]]

    proc heads(x: MySeqs, n: int): MySeqs {.noInit.}=
      discard

    proc main() =
      const iter = 10_000_000
      when defined(release):
        const report = "Release mode, time taken for " & $iter & " iterations: "
      else:
        const report = "Debug mode, time taken " & $iter & " iterations: "

      let start = cpuTime()
      for _ in 0 ..< iter:
        var x: MySeqs = (newSeq[int](10), newSeq[int](20))
        x.first[0] = 123
        let y = x.heads(3)
        discard y
      let stop = cpuTime()

      echo report & $(stop - start) & " seconds."

    main()

    ## On i5-5227U (Broadwell 2.7 Ghz turbo 3.1)
    # Release mode, time taken for 10000000 iterations: 0.921914 seconds.

    # With noInit
    # Release mode, time taken for 10000000 iterations: 0.7080679999999999 seconds.
    ```

3. **1.82x perf ratio** - Frol's completely unscientific benchmark
    ```Nim
    import random

    type Node = ref object
     # If return type is just Node, it works, but a tuple of Nodes will trigger genericReset
      x, y: int32
      left, right: Node

    proc new_node(x: int32): Node {.noinit.}=
      new result
      result.x = x
      result.y = rand(high(int32).int).int32
      result.left = nil
      result.right = nil

    proc merge(lower: Node, greater: Node): Node =
      if lower.isNil:
        return greater
      elif greater.isNil:
        return lower
      elif lower.y < greater.y:
        lower.right = merge(lower.right, greater)
        return lower
      else:
        greater.left = merge(lower, greater.left)
        return greater

    proc splitBinary(orig: Node, value: int32): (Node, Node) {.noinit.}=
      # Issue here, for return type tuple
      if orig.isNil:
        return (nil, nil)
      elif orig.x < value:
        let splitPair = splitBinary(orig.right, value)
        orig.right = splitPair[0]
        return (orig, splitPair[1])
      else:
        let splitPair = splitBinary(orig.left, value)
        orig.left = splitPair[1]
        return (splitPair[0], orig)

    proc merge3(lower: Node, equal: Node, greater: Node): Node =
      result = merge(merge(lower, equal), greater)

    proc split(orig: Node, value: int32): tuple[lower, equal, greater: Node] =
      let (lower, equalGreater) = splitBinary(orig, value)
      result.lower = lower
      (result.equal, result.greater) = splitBinary(equalGreater, value + 1)

    type Tree = ref object
      root: Node

    proc hasValue(self: Tree, x: int32): bool =
      let splited = split(self.root, x)
      result = not splited.equal.isNil
      self.root = merge3(splited.lower, splited.equal, splited.greater)

    proc insert(self: Tree, x: int32) =
      var splited = split(self.root, x)
      if splited.equal.isNil:
        splited.equal = newNode(x)
      self.root = merge3(splited.lower, splited.equal, splited.greater)

    proc erase(self: Tree, x: int32) =
      let splited = split(self.root, x)
      self.root = merge(splited.lower, splited.greater)

    proc main() =
      let tree = Tree()
      var cur = 5'i32
      var res = 0

      for i in 1 ..< 1000000:
        let a = i mod 3
        cur = (cur * 57 + 43) mod 10007
        case a:
        of 0:
          tree.insert(cur)
        of 1:
          tree.erase(cur)
        of 2:
          if tree.hasValue(cur):
            res += 1
        else:
          continue
      echo res

    when isMainModule:
      main()
    ```
    ```
    $ time build/bench
    331665

    real    0m1.222s
    user    0m1.193s
    sys     0m0.013s
    ```
    versus noinit:
    ```
    $ time build/bench
    331665

    real    0m0.672s
    user    0m0.654s
    sys     0m0.012s
    ```

4. **6x performance loss** - on #6501
    Note that I can't reproduce the initial issue on my machine on current devel (assigning result value is slower than assigning each field separately, i.e. boo slower than boo2)
    ```Nim
    import times

    type Foo = object
      val: int
      s: seq[int]

    proc newFoo(): Foo {.noInit.}=
      result.val = 1

    proc boo(): Foo {.noInit.}=
      result = newFoo()

    proc boo2(): Foo {.noInit.}=
      let a = newFoo()
      result.s = a.s
      result.val = a.val

    proc main() =
      block:
        let start = epochTime()
        for i in 0..<1000000000:
          let a = boo()
        echo "boo ", epochTime() - start

      block:
        let start = epochTime()
        for i in 0..<1000000000:
          let a = boo2()
        echo "boo2 ", epochTime() - start

    main()

    ## On i5-5227U (Broadwell 2.7 Ghz turbo 3.1)
    # boo 39.68478393554688
    # boo2 47.40634489059448

    ## With noInit
    # boo 1.9073486328125e-05 <----- optimized away by compiler
    # boo2 7.95102596282959
    ```

**Comment by Araq** [comment]
The easiest solution for now is to specialize the ``genericReset`` implementation.

**Comment by timotheecour** [comment]
related: https://github.com/nim-lang/Nim/issues/10132 shows that according to both callgrind and OSX instruments time profile, it looks like genericReset and genericResetAux are the main performance bottlenecks for nim compiler (with -d:release)

optimizing the number of times these are called would therefore improve both compile times and run times of applications

As mentioned by @arnetheduck on IRC, there are redundant calls:

> well, right now it resets in both caller and callee:

```
N_LIB_PRIVATE N_NIMCALL(void, f_aIMgZijLPElO1Hg5vcnnbA)(tyObject_X_dRG9cOkTZhLZpIMn9aGJuKpg* Result) {
        genericReset((void*)Result, (&NTI_dRG9cOkTZhLZpIMn9aGJuKpg_));
        unsureAsgnRef((void**) (&(*Result).x), copyString(((NimStringDesc*) &TM_sM5mjLRyUdt9b7mgQB49cGSg_2)));
}

N_LIB_PRIVATE N_NIMCALL(void, y_9a4WBHNDU80PDeK49bLgWWWA)(void) {
        tyArray_nHXaesL0DJZHyVS07ARPRA T1_;
        tyObject_X_dRG9cOkTZhLZpIMn9aGJuKpg T2_;
        nimZeroMem((void*)T1_, sizeof(tyArray_nHXaesL0DJZHyVS07ARPRA));
        nimZeroMem((void*)(&T2_), sizeof(tyObject_X_dRG9cOkTZhLZpIMn9aGJuKpg));
        f_aIMgZijLPElO1Hg5vcnnbA((&T2_));
        T1_[0] = copyString(T2_.x);
        echoBinSafe(T1_, 1);
}
 proc f(): X =
  result.x = "test"

proc y() =
  echo f().x
```

**Comment by Clyybber** [comment]
Fixed in #14398 afaict
